gl/vk: Enable vertex texture fetch

kd-11 opened the pull request:
Enable vertex textures.
Most of the rsx side was already done, so adding to the renderer should not be a big deal.
However, currently, the decompiler only addresses vertex texture 0.

This PR enables vertex texture fetch in GL backend only (for now)

kd-11 commented:
@raven02 Please check for regressions. The vertex texture sample works, but the texture seems to have repeat set to false (atleast for the W32_Z32_Y32_X32_FLOAT test) which confines the geometry to a small block. Maybe we shouldn't honor clamping modes?

kd-11 commented:
![image](https://cloud.githubusercontent.com/assets/15904127/18258163/7e39b4b8-73d9-11e6-8fd4-945ce91629d8.png)

Sort of working now. Some bugs are visible but they seem unrelated, I think the strange geometry over the "terrain" may be a shader effect for depth fog that isn't working properly. If you can get to the underside of the "terrain" you can see the vertex deformations are correct.

kd-11 commented:
Spurious build failure with appveyor.

raven02 commented:
Nice @kd-11

Nezarn commented:
Hatsune Miku Dreamy Theater, 2nd, and extend can now boot
http://i.imgur.com/LTAHxP1.png

ps: someone should restart appveyor, since it didn't had any build errors (at least i can't see any), and also VS builds fine

raven02 commented:
I tested a bit as well and looks okay for most games.

Appveyor build - https://ci.appveyor.com/project/raven02/rpcs3-ppr74/build/1.0.74/artifacts

kd-11 commented:
Thats good news. Will update with vulkan and dx12 support.

kd-11 commented:
Fixed opengl vertex issue, it seems rsx decode returns 0x7fffffff when restart index is 0xffffffff. For some reason bit 31 is 0 when it shouldn't be.
Added vulkan vertex textures and fix for cases where custom restart indices would be used

D3D12 vertex textures is still WIP. I'm almost certain I messed up the descriptor bindings somehow and I'm short on time to keep digging. @vlj please take a look (I get the invalid parameter error)

@raven02 please test for vulkan and ogl regressions

Some more work needs to go into the decompiler itself to get the vertex index being sampled. The dst opcode structure still contains some unknowns. I'll have to ask from those familiar with rsx internals (@vlj) or refer to nouveau docs for the ISA. If anyone has insight into this feel free to chip in. Should probably be  a 2-bit encoding in one of the Dx sections.

Nekotekina commented on rpcs3/Emu/RSX/gcm_enums.cpp:
```diff
@@ -563,6 +563,7 @@ enum
 	CELL_GCM_TEXTURE_MAX_ANISO_16 = 7,
 
 	// Texture Filter
+	CELL_GCM_TEXTURE_FILTER_NONE = 0,
```
There is no such filter.

kd-11 replied:
Vertex textures do not specify filtering. In the geforce 7 era, filtering vs textures was not supported at all. I checked this, and on the vertex_texture sample, texture filter modes are set to 0.

Nekotekina replied:
Whatever they do, CELL_GCM_TEXTURE_FILTER_NONE simply doesn't exist. We didn't invent these constants, it's confusing. And note my recent issue about gcm_enums, this problem seems relevant that the values are checked in the situation they shouldn't.

raven02 commented:
Tested for some games with both GL and Vulkan backend .It looks good.

SakataGintokiYT commented:
<img src="http://www.fotoszok.pl/upload/a0e71c7e.jpg" alt="Fotoszok.pl" />

kd-11 commented:
I'll have to revert some texture format changes on vulkan/gl.

kd-11 commented:
@SakataGintokiYT Are the vulkan/gl issues regressions? If so, @raven02 I'll need your help to figure out which texture format I botched here. I only changed swizzles for X32/16 and W32_Z32_Y32_X32 which were wrong based on inspection of shader code using the textures.

SakataGintokiYT commented:
@kd-11 
No , is not a regression on Opengl and Vulkan it was so all the time

SakataGintokiYT commented:
@kd-11 
This is not a regression

kd-11 commented:
Minor bugfixes and reworked a few areas.
Currently:

OpenGL
![image](https://cloud.githubusercontent.com/assets/15904127/18367643/e1d10268-7624-11e6-98bc-167cf088e97d.png)

Vulkan
![image](https://cloud.githubusercontent.com/assets/15904127/18367683/0edce966-7625-11e6-86dc-35032b50533d.png)

For D3D12, the shaders generated are okay, but the upload/binding code is probably where it breaks.

raven02 commented:
@SakataGintokiYT , is master also broken ? as i found some games already broken with master and not related to this vertex texture enhancement.

SakataGintokiYT commented:
@kd-11 
@raven02 
The games of naruto are broken in master build , in VK and GL Is not a regression this commit

raven02 commented:
Thanks for checking this out .It would be nice if you can cross check which commit start the breaking so we can fix the regression .

SakataGintokiYT commented:
@raven02 
Naruto games only work in DX12  .  Never work in Vulkan and OGL so it's not a regression

kd-11 commented:
I've removed dx12 part for now. The shaders and upload code is fine, but the resource binding/descriptor stuff is causing a mismatch. See https://github.com/kd-11/rpcs3/tree/dx12

Travis-Ci and appveyor are crashing before compilation starts, I'm not sure why. Should be good to go as part 1 for now as it shouldn't cause any regressions.

raven02 commented:
Need someone to restart appveyor

kd-11 commented:
Appveyor failing again during pre-compilation setup

raven02 commented:
@Nekotekina , please help to restart the appveyor.
